File: src/lib/roles.ts
import { MANAGEMENT_ROLES, TEAM_ROLES } from "./constants";

export type TeamRole = (typeof TEAM_ROLES)[number]["id"];

export function parseRoles(raw: string | null | undefined): string[] {
  return [...new Set((raw || "")
    .split(",")
    .map((item) => item.trim().toUpperCase())
    .filter((item) => TEAM_ROLES.some((role) => role.id === item)))];
}

export function joinRoles(roles: string[]) {
  const unique = parseRoles(roles.join(","));
  return unique.length ? unique.join(",") : "OTHER";
}

export function hasManagementRole(roles: string[] | string) {
  const parsed = Array.isArray(roles) ? roles : parseRoles(roles);
  return parsed.some((role) => (MANAGEMENT_ROLES as readonly string[]).includes(role));
}

export function isProviderRole(roles: string[] | string) {
  const parsed = Array.isArray(roles) ? roles : parseRoles(roles);
  return parsed.some((role) => role !== "OWNER" && role !== "MANAGER") || parsed.includes("OWNER");
}

export function roleLabel(id: string, lang: "ar" | "en") {
  const row = TEAM_ROLES.find((role) => role.id === id);
  return row ? row[lang] : id;
}

export function rolesFromSpecialty(specialty: string) {
  const parts = specialty.split(",").map((item) => item.trim());
  const roles = new Set<string>(["OWNER"]);
  for (const part of parts) {
    if (part === "hair") roles.add("HAIRSTYLIST");
    else if (part === "nails") roles.add("NAIL_ARTIST");
    else if (part === "makeup" || part === "henna" || part === "skincare" || part === "photo" || !part) {
      roles.add("MAKEUP_ARTIST");
    } else roles.add("OTHER");
  }
  if (roles.size === 1) roles.add("MAKEUP_ARTIST");
  return joinRoles([...roles]);
}

export function defaultCapacityForRoles(roles: string[] | string) {
  const parsed = Array.isArray(roles) ? roles : parseRoles(roles);
  if (parsed.includes("HAIRSTYLIST") && !parsed.includes("MAKEUP_ARTIST")) return 5;
  return 4;
}

export function rolesForServiceKind(kind: string) {
  switch (kind) {
    case "hair":
      return ["HAIRSTYLIST"];
    case "nails":
      return ["NAIL_ARTIST"];
    case "henna":
      return ["MAKEUP_ARTIST", "OTHER"];
    case "other":
      return ["MAKEUP_ARTIST", "HAIRSTYLIST", "NAIL_ARTIST", "LASH_ARTIST", "OTHER"];
    default:
      return ["MAKEUP_ARTIST"];
  }
}

export function memberMatchesServiceKind(roles: string[] | string, kind: string) {
  const parsed = Array.isArray(roles) ? roles : parseRoles(roles);
  return rolesForServiceKind(kind).some((role) => parsed.includes(role));
}

/** Salon type, or more than one active member (solo expert who has started hiring). */
export function isTeamBusiness(businessType?: string | null, activeMemberCount = 0) {
  return businessType === "salon" || activeMemberCount > 1;
}

File: backend/bridey_api/roles.py
from __future__ import annotations

from bridey_api.constants import MANAGEMENT_ROLES, TEAM_ROLES


def parse_roles(raw: str | list[str] | None) -> list[str]:
    if isinstance(raw, list):
        parts = raw
    else:
        parts = (raw or "").split(",")
    seen: list[str] = []
    for item in parts:
        role = item.strip().upper()
        if role in TEAM_ROLES and role not in seen:
            seen.append(role)
    return seen


def join_roles(roles: list[str]) -> str:
    unique = parse_roles(",".join(roles))
    return ",".join(unique) if unique else "OTHER"


def has_management_role(roles: list[str] | str) -> bool:
    parsed = roles if isinstance(roles, list) else parse_roles(roles)
    return any(role in MANAGEMENT_ROLES for role in parsed)


def roles_from_specialty(specialty: str) -> str:
    parts = [item.strip() for item in specialty.split(",")]
    roles = {"OWNER"}
    for part in parts:
        if part == "hair":
            roles.add("HAIRSTYLIST")
        elif part == "nails":
            roles.add("NAIL_ARTIST")
        elif part in {"makeup", "henna", "skincare", "photo", ""}:
            roles.add("MAKEUP_ARTIST")
        else:
            roles.add("OTHER")
    if len(roles) == 1:
        roles.add("MAKEUP_ARTIST")
    return join_roles(list(roles))


def default_capacity_for_roles(roles: list[str] | str) -> int:
    parsed = roles if isinstance(roles, list) else parse_roles(roles)
    if "HAIRSTYLIST" in parsed and "MAKEUP_ARTIST" not in parsed:
        return 5
    return 4


def roles_for_service_kind(kind: str) -> list[str]:
    if kind == "hair":
        return ["HAIRSTYLIST"]
    if kind == "nails":
        return ["NAIL_ARTIST"]
    if kind == "henna":
        return ["MAKEUP_ARTIST", "OTHER"]
    if kind == "other":
        return ["MAKEUP_ARTIST", "HAIRSTYLIST", "NAIL_ARTIST", "LASH_ARTIST", "OTHER"]
    return ["MAKEUP_ARTIST"]


def member_matches_service_kind(roles: list[str] | str, kind: str) -> bool:
    parsed = roles if isinstance(roles, list) else parse_roles(roles)
    return any(role in parsed for role in roles_for_service_kind(kind))
